plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftSplitManager.java

```diff
@@ -0,0 +1,202 @@
+        implements ConnectorSplitManager
+{
+    private static final Logger log = Logger.get(RedshiftSplitManager.class);
+
+    private final JdbcClient jdbcClient;
+    private final QueryBuilder queryBuilder;
+    private final RemoteQueryModifier queryModifier;
+    private final JdbcSplitManager jdbcSplitManager;
+    private final Optional<String> unloadLocation;
+    private final Optional<String> unloadAuthorization;
+    private final ExecutorService executor;
+    private final TrinoFileSystemFactory fileSystemFactory;
+
+    @Inject
+    public RedshiftSplitManager(
+            JdbcClient jdbcClient,
+            QueryBuilder queryBuilder,
+            RemoteQueryModifier queryModifier,
+            JdbcSplitManager jdbcSplitManager,
+            RedshiftConfig redshiftConfig,
+            @FileSystemS3 TrinoFileSystemFactory fileSystemFactory,
+            @ForRecordCursor ExecutorService executor)
+    {
+        this.jdbcClient = requireNonNull(jdbcClient, "jdbcClient is null");
+        this.queryBuilder = requireNonNull(queryBuilder, "queryBuilder is null");
+        this.queryModifier = requireNonNull(queryModifier, "queryModifier is null");
+        this.jdbcSplitManager = requireNonNull(jdbcSplitManager, "jdbcSplitManager is null");
+        this.unloadLocation = redshiftConfig.getUnloadLocation();
+        this.unloadAuthorization = redshiftConfig.getUnloadIamRole();
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        this.executor = requireNonNull(executor, "executor is null");
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorTableHandle table, DynamicFilter dynamicFilter, Constraint constraint)
+    {
+        if (table instanceof JdbcProcedureHandle) {
+            return jdbcSplitManager.getSplits(transaction, session, table, dynamicFilter, constraint);
+        }
+        ConnectorSplitSource fallbackSplitSource = new FixedSplitSource(new JdbcSplit(Optional.empty()));
+        if (!isUnloadEnabled(session)) {
+            return fallbackSplitSource;
+        }
+        JdbcTableHandle jdbcTable = (JdbcTableHandle) table;
+        JdbcTableHandle jdbcTableHandle = dynamicFilteringEnabled(session) ? jdbcTable.intersectedWithConstraint(dynamicFilter.getCurrentPredicate()) : jdbcTable;
+        List<JdbcColumnHandle> columns = jdbcTableHandle.getColumns()
+                .orElseGet(() -> jdbcClient.getColumns(
+                        session,
+                        jdbcTableHandle.getRequiredNamedRelation().getSchemaTableName(),
+                        jdbcTableHandle.getRequiredNamedRelation().getRemoteTableName()));
+
+        if (!isUnloadSupported(jdbcTable, columns)) {
+            log.debug("Unsupported query shape detected. Falling back to using JDBC");
+            return fallbackSplitSource;
+        }
+
+        Connection connection;
```

Comment:
Who closes this connection?

Reply:
Connection is closed by `RedshiftUnloadSplitSource#close`
But I think it will leak when there is an exception here